### PR TITLE
cdp: improved routing and removed useless warning/errors in authstatemachine

### PR DESCRIPTION
**Pull request by vingarzan**
#### Pre-Submission Checklist

- [x] Commit message has the format required by CONTRIBUTING guide
- [x] Commits are split per component (core, individual modules, libs, utils, ...)
- [x] Each component has a single commit (if not, squash them into one commit)
- [x] No commits to README files for modules (changes must be done to docbook files
in `doc/` subfolder, the README file is autogenerated)

#### Type Of Change
- [x] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)

#### Checklist:

- [ ] PR should be backported to stable branches
- [x] Tested changes locally
- [ ] Related to issue #XXXX (replace XXXX with an open issue number)

#### Description

Diameter requests could contain Application-Ids in multiple AVPs. This fix iterates through them, instead of giving up if the first one does not match. This solves some issues when peers don't always declare support for vendor-specific/non-specific application ids, so routing should be easier.

Also fixed:
- auth state machine being too verbose on regular messages being received or sent while in the Open state
- add Destination-Host from auth state to requests (next-hop immediate routing)

### CODE REVIEW

File: src/modules/cdp/authstatemachine.c [hunk 1]
```diff
@@ -394,41 +394,54 @@ int auth_client_statefull_sm_process(
...
+					LM_DBG("auth_client_statefull_sm_process(): Received "
+						   "request while in state %s\n",
+							auth_states[x->state]);
+					break;
+
 				case AUTH_EV_RECV_ANS_SUCCESS:
 					x->state = AUTH_ST_OPEN;
 					update_auth_session_timers(x, msg);
-					//LM_INFO("state machine: i was in open and i am going to open\n");
+					//LM_INFO("state machine: I was in open and I am going to open\n");
```
Comment (herlesupreeth) [inline src/modules/cdp/authstatemachine.c:413]: I thin we can remove this commented code

File: src/modules/cdp/authstatemachine.c [hunk 2]
```diff
@@ -394,41 +394,54 @@ int auth_client_statefull_sm_process(
...
 					break;
 
 				case AUTH_EV_RECV_ANS_UNSUCCESS:
 					x->state = AUTH_ST_DISCON;
-					//LM_INFO("state machine: i was in open and i am going to discon\n");
+					//LM_INFO("state machine: I was in open and I am going to discon\n");
```
Comment (herlesupreeth) [inline src/modules/cdp/authstatemachine.c:418]: same as above. remove the commented code

File: src/modules/cdp/authstatemachine.c [hunk 3]
```diff
@@ -394,41 +394,54 @@ int auth_client_statefull_sm_process(
...
 					break;
 
 				case AUTH_EV_SESSION_TIMEOUT:
 				case AUTH_EV_SERVICE_TERMINATED:
 				case AUTH_EV_SESSION_GRACE_TIMEOUT:
 					x->state = AUTH_ST_DISCON;
-					//LM_INFO("state machine: i was in open and i am going to discon\n");
+					//LM_INFO("state machine: I was in open and I am going to discon\n");
```
Comment (herlesupreeth) [inline src/modules/cdp/authstatemachine.c:425]: same as above

File: src/modules/cdp/authstatemachine.c [hunk 4]
```diff
@@ -394,41 +394,54 @@ int auth_client_statefull_sm_process(
...
 
 					Send_STR(s, msg);
 					break;
 
 				case AUTH_EV_SEND_ASA_SUCCESS:
 					x->state = AUTH_ST_DISCON;
-					//LM_INFO("state machine: i was in open and i am going to discon\n");
+					//LM_INFO("state machine: I was in open and I am going to discon\n");
```
Comment (herlesupreeth) [inline src/modules/cdp/authstatemachine.c:432]: same as above

File: src/modules/cdp/authstatemachine.c [hunk 5]
```diff
@@ -394,41 +394,54 @@ int auth_client_statefull_sm_process(
...
 					Send_STR(s, msg);
 					break;
 
 				case AUTH_EV_SEND_ASA_UNSUCCESS:
 					x->state = AUTH_ST_OPEN;
 					update_auth_session_timers(x, msg);
-					//LM_INFO("state machine: i was in open and i am going to open\n");
+					//LM_INFO("state machine: I was in open and I am going to open\n");
```
Comment (herlesupreeth) [inline src/modules/cdp/authstatemachine.c:439]: remove commented code

**Review by herlesupreeth — CHANGES_REQUESTED** [review]

File: src/modules/cdp/routing.c [hunk 6]
```diff
@@ -200,102 +251,101 @@ peer *get_routing_peer(cdp_session_t *cdp_session, AAAMessage *m)
...
+		p = get_peer_by_fqdn(&destination_host);
+		if(p && (p->state == I_Open || p->state == R_Open)
+				&& peer_handles_application(p, app_id, vendor_id)) {
+			p->last_selected = time(NULL);
+			return p;
+		}
+		/* the destination host peer is not connected at the moment, try a normal route then */
+	}
+
+	if(!config->r_table) {
```
Comment (herlesupreeth) [inline src/modules/cdp/routing.c:272]: cant we remove this check here as its already done in `get_peer_by_fqdn` ??

File: src/modules/cdp/routing.c [hunk 7]
```diff
@@ -200,102 +251,101 @@ peer *get_routing_peer(cdp_session_t *cdp_session, AAAMessage *m)
...
 	peer *p;
-	routing_realm *rr;
 	int app_id = 0, vendor_id = 0;
 
+	// 1. Destination-Host routing to adjacent peer
+	avp = AAAFindMatchingAVP(m, 0, AVP_Destination_Host, 0, AAA_FORWARD_SEARCH);
+	if(avp)
+		destination_host = avp->data;
+
+	if(destination_host.len) {
```
Comment (herlesupreeth) [inline src/modules/cdp/routing.c:261]: cant we remove this check here as its already done in `get_peer_by_fqdn` ??

**Comment by herlesupreeth** [comment]
Thanks @vingarzan. Looks good to me.

**Comment by vingarzan** [comment]
> Thanks @vingarzan. Looks good to me.

Thank you also for the review! I have removed the commented out logs.

**Review by herlesupreeth — APPROVED** [review]

**Pull request merged**